quaternions.scad

```diff
@@ -327,4 +447,216 @@ function Qrot(q,p) =
+//   #sphere(r=80);
+// Example(3D): as a module
+//   a = QuatY(5);
+//   Q_Rotation_path(a, 10)
+//      right(80) cube([10,10,1]);
+//   #sphere(r=80);
+function Q_Rotation_path(q1, n=1, q2) =
+  assert( Q_is_quat(q1) && (is_undef(q2) || Q_is_quat(q2) ), "Invalid quaternion(s)" )
+  assert( is_num(0*n) && n>=1 && n==floor(n), "Invalid integer" )
+  assert( is_undef(q2) || ! approx(norm(q1+q2),0), "Quaternions cannot be opposed" )
+  is_undef(q2)
+  ? [for( i=0, dR=Q_Matrix4(q1), R=dR; i<=n; i=i+1, R=dR*R ) R]
+  : let( q2 = Q_Normalize( q1*q2<0 ? -q2: q2 ) )
+    let( dq = Q_pow( Q_Mul( q2, Q_Inverse(q1) ), 1/n ),
+         dR = Q_Matrix4(dq) )
+    [for( i=0, R=Q_Matrix4(q1); i<=n; i=i+1, R=dR*R ) R];
+
+module Q_Rotation_path(q1, n=1, q2) {
+  for(Mi=Q_Rotation_path(q1, n, q2))
+    multmatrix(Mi)
+      children();
+}
+
+
+// Function: Q_Nlerp()
+// Usage:
+//   q = Q_Nlerp(q1, q2, u);
+// Description:
+//   Returns a quaternion that is a normalized linear interpolation between two quaternions
+//   when u is a number.
+//   If u is a list of numbers, computes the interpolations for each value in the
+//   list and returns the interpolated quaternions in a list.
+//   The input quaternions don't need to be normalized.
+// Arguments:
+//   q1 = The first quaternion. (u=0)
+//   q2 = The second quaternion. (u=1)
+//   u  = A value (or a list of values), between 0 and 1, of the proportion(s) of each quaternion in the interpolation. 
+// Example(3D): Giving `u` as a Scalar
+//   a = QuatY(-135);
+//   b = QuatXYZ([0,-30,30]);
+//   for (u=[0:0.1:1])
+//       Qrot(Q_Nlerp(a, b, u))
+//           right(80) cube([10,10,1]);
+//   #sphere(r=80);
+// Example(3D): Giving `u` as a Range
+//   a = QuatZ(-135);
+//   b = QuatXYZ([90,0,-45]);
+//   for (q = Q_Nlerp(a, b, [0:0.1:1]))
+//       Qrot(q) right(80) cube([10,10,1]);
+//   #sphere(r=80);
+function Q_Nlerp(q1,q2,u) =
+  assert(Q_is_quat(q1) && Q_is_quat(q2), "Invalid quaternion(s)" )
+  assert( ! approx(norm(q1+q2),0), "Quaternions cannot be opposed" )
+  assert(is_num(0*u) || is_range(u) || (is_list(u) && is_num(0*u*u)) ,
+         "Invalid interpolation coefficient(s)" )
+  let( q1  = Q_Normalize(q1),
+       q2  = Q_Normalize(q2) )
+  !is_num(u)
+  ? [for (ui=u) _Qnorm((1-ui)*q1 + ui*q2 ) ]
+  : _Qnorm((1-u)*q1 + u*q2 );
+
+
+// Function: Q_Squad()
+// Usage:
+//   qn = Q_Squad(q1,q2,q3,q4,u);
+// Description:
+//   Returns a quaternion that is a cubic spherical interpolation of the quaternions  
+//   q1 and q4 taking the other two quaternions, q2 and q3, as parameter of a cubic 
+//   on the sphere similar to the control points of a Bezier curve.
+//   If u is a number, usually between 0 and 1, returns the quaternion that results
+//   from the interpolation. 
+//   If u is a list of numbers, computes the interpolations for each value in the
+//   list and returns the interpolated quaternions in a list.
+//   The input quaternions don't need to be normalized.

```

Comment:
Bezier-like quaternion interpolation?  Mind blown!

Reply:
Yes, it is. I can't see myself using this function but for sake of completeness I include it. The same with Q_exp and Q_ln although those are useful to compare with alternative expressions. For instance, quaternion power is usually defined based in exp and ln but I was successful in express a faster alternative without them. 

Do you believe there are things like Squad based on B-splines?

Reply:
I once proposed support for a full NURBS extension on Beziers, but we decided that would make for a rather complicated interface for minimal extra functionality.